src/schemas.jl

```diff
@@ -778,8 +781,26 @@ macro version(record_type, declared_fields_block=nothing)
 
     # parse `declared_fields_block`
     declared_field_statements = Any[]
+    declared_constraint_statements = Any[]
     if declared_fields_block isa Expr && declared_fields_block.head == :block && !isempty(declared_fields_block.args)
-        declared_field_statements = [f for f in declared_fields_block.args if !(f isa LineNumberNode)]
+        for f in declared_fields_block.args
+            if f isa LineNumberNode
+                continue
+            elseif f isa Expr && f.head === :macrocall && f.args[1] === Symbol("@check")
+                # Expand `@check` macro here so that we can reliably see the location of
+                # the user define `@check` when it fails. Ideally `Meta.replace_sourceloc!`
+                # would do this for us.
+                constraint_expr = Base.macroexpand(Legolas, f)
+                if f.args[2] isa LineNumberNode
+                    constraint_expr = Expr(:block, f.args[2], constraint_expr)
+                end
```

Comment:
this is pretty opaque to me, add a comment explaining why this is necessary?

Reply:
we don't want to just strip out the line number nodes right, since that's what will give the nice error w/ teh correct definition location when the check fails, right?

Reply:
ah wait I think I see: we're creating a new `begin...end` with the line number node in the macrocall manually spliced in (if there is one).

Reply:
I've updated the comment to something which I think is clearer